utils/clients/gemini/client.go

```diff
@@ -0,0 +1,279 @@
+	return BalancesPayload{
+		Request: "/v1/balances",
+		Nonce:   nonce(),
+		Account: account,
+	}
+}
+
+// PayoutResult contains details about a newly created or fetched issuer
+type PayoutResult struct {
+	Result      string           `json:"result"` // OK or Error
+	TxRef       string           `json:"tx_ref"`
+	Amount      *decimal.Decimal `json:"amount"`
+	Currency    *string          `json:"currency"`
+	Destination *string          `json:"destination"`
+	Status      *string          `json:"status"`
+	Reason      *string          `json:"reason"`
+}
+
+// Balance holds balance info
+type Balance struct {
+	Type                   string          `json:"type"`
+	Currency               string          `json:"currency"`
+	Amount                 decimal.Decimal `json:"amount"`
+	Available              decimal.Decimal `json:"available"`
+	AvailableForWithdrawal decimal.Decimal `json:"availableForWithdrawal"`
+}
+
+// Account holds account info
+type Account struct {
+	Name           string `json:"name"`
+	Class          string `json:"account"`
+	Type           string `json:"type"`
+	CounterpartyID string `json:"counterparty_id"`
+	CreatedAt      int64  `json:"created"`
+}
+
+// GenerateLog creates a log
+func (pr PayoutResult) GenerateLog() string {
+	if pr.Result == "OK" {
+		return ""
+	}
+	return strings.Join([]string{pr.Result, pr.TxRef, *pr.Status, *pr.Reason}, ": ")
+}
+
+// Client abstracts over the underlying client
+type Client interface {
+	// FetchAccountList requests account information to scope future requests
+	FetchAccountList(ctx context.Context, APIKey string, signer cryptography.HMACKey, payload string) (*[]Account, error)
+	// FetchBalances requests balance information for a given account
+	FetchBalances(ctx context.Context, APIKey string, signer cryptography.HMACKey, payload string) (*[]Balance, error)
+	// UploadBulkPayout posts a signed bulk layout to gemini
+	UploadBulkPayout(ctx context.Context, APIKey string, signer cryptography.HMACKey, payload string) (*[]PayoutResult, error)
+}
+
+// HTTPClient wraps http.Client for interacting with the cbr server
+type HTTPClient struct {
+	client *clients.SimpleHTTPClient
+}
+
+// New returns a new HTTPClient, retrieving the base URL from the environment
+func New() (Client, error) {
+	serverEnvKey := "GEMINI_SERVER"
+	serverURL := os.Getenv("GEMINI_SERVER")
+	if len(serverURL) == 0 {
+		return nil, errors.New(serverEnvKey + " was empty")
+	}
+	client, err := clients.New(serverURL, os.Getenv("GEMINI_TOKEN"))
+	if err != nil {
+		return nil, err
+	}
+	return NewClientWithPrometheus(&HTTPClient{client}, "gemini_client"), err
+}
+
+func setPrivateRequestHeaders(req *http.Request, APIKey string, signer cryptography.HMACKey, payload string) error {
+	signature, err := signer.HMACSha384([]byte(payload))
+	if err != nil {
+		return err
+	}
+	req.Header.Set("Content-Type", "text/plain")
+	req.Header.Set("Content-Length", "0")
+	req.Header.Set("X-GEMINI-PAYLOAD", payload)
```

Comment:
Very strange API, takes the request payload in a header, and no request body?

Reply:
yeah, just how it is unfortunately